Integrations/Forescout/Forescout.py

```diff
@@ -0,0 +1,678 @@
+            }
+            formatted_rules.append(formatted_rule)
+        formatted_policy['Rule'] = formatted_rules
+        formatted_policies.append(formatted_policy)
+    return formatted_policies
+
+
+def create_web_api_headers() -> Dict:
+    """
+    Update JWT if it has expired and return headers object that formats to Forescout Web API expectations
+
+    Returns
+    -------
+    dict
+        Headers object for the Forescout Web API calls
+    """
+    login()
+    headers = {
+        'Authorization': WEB_AUTH,
+        'Accept': 'application/hal+json'
+    }
+    return headers
+
+
+def login():
+    global LAST_JWT_FETCH
+    global WEB_AUTH
+    if not LAST_JWT_FETCH or datetime.now(timezone.utc) >= LAST_JWT_FETCH + JWT_VALIDITY_TIME:
+        url_suffix = '/api/login'
+        headers = {'Content-Type': 'application/x-www-form-urlencoded'}
+        params = {'username': USERNAME, 'password': PASSWORD}
+        response = http_request('POST', url_suffix, headers=headers, params=params, resp_type='response')
+        fetch_time = parsedate(response.headers.get('Date', ''))
+        WEB_AUTH = response.text
+        LAST_JWT_FETCH = fetch_time
+
+
+def http_request(method: str, url_suffix: str, full_url: str = None, headers: Dict = None, auth: Tuple = None,
+                 params: Dict = None, data: Dict = None, files: Dict = None, resp_type: str = 'json') -> Any:
+    """
+    A wrapper for requests lib to send our requests and handle requests
+    and responses better
+
+    Parameters
+    ----------
+    method : str
+        HTTP method, e.g. 'GET', 'POST' ... etc.
+    url_suffix : str
+        API endpoint.
+    full_url : str
+        Bypasses the use of BASE_URL + url_suffix. Useful if there is a need to
+        make a request to an address outside of the scope of the integration
+        API.
+    headers : dict
+        Headers to send in the request.
+    auth : tuple
+        Auth tuple to enable Basic/Digest/Custom HTTP Auth.
+    params : dict
+        URL parameters.
+    data : dict
+        Data to be sent in a 'POST' request.
+    files : dict
+        File data to be sent in a 'POST' request.
+    resp_type : str
+        Determines what to return from having made the HTTP request. The default
+        is 'json'. Other options are 'text', 'content' or 'response' if the user
+        would like the full response object returned.
+
+    Returns
+    -------
+    dict | str | bytes | obj
+        Response JSON from having made the request.
+    """
+    try:
+        address = full_url if full_url else BASE_URL + url_suffix
+        res = requests.request(
+            method,
+            address,
+            verify=USE_SSL,
+            params=params,
+            data=data,
+            files=files,
+            headers=headers,
+            auth=auth
+        )
+
+        # Handle error responses gracefully
+        if res.status_code not in {200, 201, 304}:
```

Comment:
whats 201 and 304 ?

Reply:
> If the queried data has not changed in the Forescout platform since the
last request, clients receive an HTTP 304 (Not Modified) return message.

201 is leftover from previous integration I copied the basic version of `http_request` from and I just left it. I can remove it if you'd like.